### PR TITLE
Parallel all the applications within cloudbuild test

### PR DESCRIPTION
**Pull request by chiayi**
Parallel the application builds. Run Jupyter/Rag/Ray at the same time in difference namespaces.

### CODE REVIEW

**Comment by chiayi** [comment]
Jupyter seems to be running into some issue with waiting for pods to be ready. Currently happening when rag is being installed, although I'm unsure if it's related and can't seem to repro locally.

**Comment by chiayi** [comment]
Error mentioned in this comment: https://github.com/GoogleCloudPlatform/ai-on-gke/pull/620#issuecomment-2080202976 seems to be inconsistent. Another issue that seems to occur with cloud-build is Jupyter unable to destroy properly.

**Comment by chiayi** [comment]
Cloud-build was failing due to the port for Rag being changed. It was changed to prevent Jupyter/Rag/Ray sharing the same port. Changing back to see if it succeeds.

**Comment by chiayi** [comment]
Build takes around 50 mins.

**Comment by andrewsykim** [comment]
The build time for `pr-review-trigger` is 52m, I was expecting a larger improvement from this change :thinking:

File: cloudbuild.yaml [hunk 1]
```diff
@@ -91,16 +91,16 @@ steps:
         -var=project_id=$PROJECT_ID \
         -var=cluster_name=ml-$SHORT_SHA-$_PR_NUMBER-$_BUILD_ID-cluster \
         -var=cluster_location=$_REGION \
-        -var=kubernetes_namespace=ml-$SHORT_SHA-$_BUILD_ID \
+        -var=kubernetes_namespace=ml-$SHORT_SHA-$_BUILD_ID-ray \
```
Comment (andrewsykim) [inline cloudbuild.yaml:94]: Changing the namespace allows the steps to run independant from each other, but do we need additional changes to run the steps in parallel? I couldn't tell from cloudbuild logs if the steps are starting in parallel

File: cloudbuild.yaml [hunk 2]
```diff
@@ -150,22 +150,23 @@ steps:
...
+        kubectl port-forward -n ml-$SHORT_SHA-$_BUILD_ID-jupyter service/proxy-public 9442:80 &
         # Wait port-forwarding to take its place
         sleep 5s
 
         cd /workspace/modules/jupyter/tests
-        python3 test_hub.py "127.0.0.1:9443" $_AUTOPILOT_CLUSTER
+        python3 test_hub.py "127.0.0.1:9442" $_AUTOPILOT_CLUSTER
         echo "pass" > /workspace/jupyterhub_test_result.txt
     allowFailure: true
+    waitFor: ['create gke cluster']
```
Comment (andrewsykim) [inline cloudbuild.yaml:169]: Update 'test jupyterhub' too?

File: cloudbuild.yaml (hunk 1 above)
Comment (andrewsykim) [reply cloudbuild.yaml:94]: Oh I see, I guess just changing the `waitFor` in the test steps makes it parallel

File: cloudbuild.yaml (hunk 2 above)
Comment (andrewsykim) [reply cloudbuild.yaml:169]: This is weird, how was this working before without a `waitFor`?

File: cloudbuild.yaml (hunk 1 above)
Comment (chiayi) [reply cloudbuild.yaml:94]: Yes, changing the `waitFor` explicitly will tell it to run right after the steps within `waitFor`

File: cloudbuild.yaml (hunk 2 above)
Comment (chiayi) [reply cloudbuild.yaml:169]: So if there is no explicit `waitFor` it will just wait for all previous steps to finish before running.

**Review by yiyinglovecoding — APPROVED** [review]
LGTM

File: cloudbuild.yaml (hunk 2 above)
Comment (chiayi) [reply cloudbuild.yaml:169]: For more info: https://cloud.google.com/build/docs/configuring-builds/configure-build-step-order#build_step_order_and_dependencies

**Comment by chiayi** [comment]
Looking at all the past runs, the cleanup would sometimes not completely succeed (all from standard cluster). Going to run one more time to see if it occurs.

**Comment by andrewsykim** [comment]
> Looking at all the past runs, the cleanup would sometimes not completely succeed (all from standard cluster). Going to run one more time to see if it occurs.

I don't think clean up ever succeeded fwiw, there's some known bugs there

**Comment by chiayi** [comment]
> > Looking at all the past runs, the cleanup would sometimes not completely succeed (all from standard cluster). Going to run one more time to see if it occurs.
> 
> I don't think clean up ever succeeded fwiw, there's some known bugs there

Oh I was specifically talking about the Jupyterhub cleanup. 
success: https://pantheon.corp.google.com/cloud-build/builds;region=us-central1/6e09787a-1d3e-4509-8c42-fe4cb3d0b0d4?e=-13802955&mods=prod_coliseum&project=gke-ai-eco-dev and https://pantheon.corp.google.com/cloud-build/builds;region=us-central1/fefbdae1-a82e-4b2a-884e-f2a8a5bc5cc2?e=-13802955&mods=prod_coliseum&project=gke-ai-eco-dev. 

fail: https://pantheon.corp.google.com/cloud-build/builds;region=us-central1/a3b49edc-e6a9-459a-9792-f017cebe9d87?e=-13802955&mods=prod_coliseum&project=gke-ai-eco-dev

**Comment by chiayi** [comment]
And I guess the latest run as well: https://pantheon.corp.google.com/cloud-build/builds;region=us-central1/110ac62c-99c9-4a8c-b4b5-4dcca73193c5;step=8?e=-13802955&mods=prod_coliseum&project=gke-ai-eco-dev

**Comment by andrewsykim** [comment]
> Oh I was specifically talking about the Jupyterhub cleanup.

Oh my bad, yeah this seems like a new failure

**Comment by chiayi** [comment]
I couldn't repro on local. Printing kubectl events to see if there is anything there.

File: cloudbuild.yaml [hunk 3]
```diff
@@ -150,22 +150,25 @@ steps:
...
         -var=gcs_bucket=gke-aieco-jupyter-$SHORT_SHA-$_BUILD_ID \
         -auto-approve -no-color -lock=false
         echo "pass" > /workspace/jupyterhub_tf_result.txt
 
-        kubectl wait --all pods -n ml-$SHORT_SHA-$_BUILD_ID --for=condition=Ready --timeout=1200s
-        kubectl get services -n ml-$SHORT_SHA-$_BUILD_ID
-        kubectl port-forward -n ml-$SHORT_SHA-$_BUILD_ID service/proxy-public 9443:80 &
+        kubectl get pods -n ml-$SHORT_SHA-$_BUILD_ID-jupyter
+        kubectl wait --all pods -n ml-$SHORT_SHA-$_BUILD_ID-jupyter --for=condition=Ready --timeout=1800s
+        kubectl events -n ml-$SHORT_SHA-$_BUILD_ID-jupyter
```
Comment (andrewsykim) [inline cloudbuild.yaml:161]: What the reason for this command?

Comment (andrewsykim) [reply cloudbuild.yaml:161]: please remove it  before merging if it's not needed

File: cloudbuild.yaml [hunk 4]
```diff
@@ -150,22 +150,25 @@ steps:
...
+        -var=kubernetes_namespace=ml-$SHORT_SHA-$_BUILD_ID-jupyter \
         -var=workload_identity_service_account=jupyter-sa-$SHORT_SHA-$_BUILD_ID \
         -var=gcs_bucket=gke-aieco-jupyter-$SHORT_SHA-$_BUILD_ID \
         -auto-approve -no-color -lock=false
         echo "pass" > /workspace/jupyterhub_tf_result.txt
 
-        kubectl wait --all pods -n ml-$SHORT_SHA-$_BUILD_ID --for=condition=Ready --timeout=1200s
-        kubectl get services -n ml-$SHORT_SHA-$_BUILD_ID
-        kubectl port-forward -n ml-$SHORT_SHA-$_BUILD_ID service/proxy-public 9443:80 &
+        kubectl get pods -n ml-$SHORT_SHA-$_BUILD_ID-jupyter
```
Comment (andrewsykim) [inline cloudbuild.yaml:159]: If  you're running kubectl get pods for debugging purposes, it should come after the `kubectl wait --all pods` below

**Review by andrewsykim — APPROVED** [review]

File: cloudbuild.yaml [hunk 5]
```diff
@@ -150,22 +150,25 @@ steps:
...
+        -auto-approve -no-color 
         echo "pass" > /workspace/jupyterhub_tf_result.txt
 
-        kubectl wait --all pods -n ml-$SHORT_SHA-$_BUILD_ID --for=condition=Ready --timeout=1200s
-        kubectl get services -n ml-$SHORT_SHA-$_BUILD_ID
-        kubectl port-forward -n ml-$SHORT_SHA-$_BUILD_ID service/proxy-public 9443:80 &
+        kubectl wait --all pods -n ml-$SHORT_SHA-$_BUILD_ID-jupyter --for=condition=Ready --timeout=1800s
+        kubectl get pods -n ml-$SHORT_SHA-$_BUILD_ID-jupyter
+        kubectl events -n ml-$SHORT_SHA-$_BUILD_ID-jupyter
+        kubectl get services -n ml-$SHORT_SHA-$_BUILD_ID-jupyter
```
Comment (andrewsykim) [inline cloudbuild.yaml:162]: Remove this command too

**Comment by chiayi** [comment]
RAG frontend isn't coming up with error: ` Waiting for rollout to finish: 3 replicas wanted; 2 replicas Ready`

**Comment by chiayi** [comment]
Cleaned up all the unnecessary `kubectl` commands. Will also try test again.

**Comment by chiayi** [comment]
Same error 

> RAG frontend isn't coming up with error: ` Waiting for rollout to finish: 3 replicas wanted; 2 replicas Ready`

Adding terraform debug and kubectl event

**Comment by chiayi** [comment]
Cleaned up the rest of the `kubectl events/get pods` commands that was used for debugging.

**Comment by chiayi** [comment]
Going to run again, the test failed due to cluster not being able to be created properly.

**Comment by chiayi** [comment]
I'm noticing that the frontend 3rd replica takes a bit longer to start up since the image takes bit longer to pull. I'm extending the timeout for create of the frontend to address this. As for why image pulling is taking longer, it will require a deeper dive into why it takes an avg of ~7m more for the 3rd replica. This is only occurring for standard.

**Comment by chiayi** [comment]
I am now confused. Error is pretty inconsistent for some reason.

File: applications/rag/frontend/main.tf [hunk 6]
```diff
@@ -190,4 +190,9 @@ resource "kubernetes_deployment" "rag_frontend_deployment" {
       }
     }
   }
+
+  timeouts {
+    create = "30m"
```
Comment (andrewsykim) [inline applications/rag/frontend/main.tf:195]: Consider experimenting with larger node pools before increasing this timeout.

Comment (andrewsykim) [reply applications/rag/frontend/main.tf:195]: (larger node pools for the tests only to be clear)

**Comment by chiayi** [comment]
Ran into network issue while autopilot cluster did not.

File: cloudbuild.yaml [hunk 7]
```diff
@@ -65,7 +65,9 @@ steps:
         -var=cluster_name=ml-$SHORT_SHA-$_PR_NUMBER-$_BUILD_ID-cluster \
         -var=autopilot_cluster=$_AUTOPILOT_CLUSTER \
         -var=cluster_location=$_REGION \
-        -auto-approve -no-color -lock=false
+        -var='cpu_pools=[{initial_node_count=2,name="cpu-pool",machine_type="n1-standard-16",autoscaling=true,min_count=1,max_count=3,disk_size_gb=100,disk_type="pd-standard",}]' \
```
Comment (andrewsykim) [inline cloudbuild.yaml:68]: We have dedicated tfvars files for tests, slightly cleaner to just add this change directly into https://github.com/GoogleCloudPlatform/ai-on-gke/blob/main/infrastructure/tfvars_tests/standard-gke-public.platform.tfvars

File: cloudbuild.yaml [hunk 8]
```diff
@@ -219,22 +220,22 @@ steps:
...
+        -var=jupyter_service_account=jupyter-sa-4-rag-$SHORT_SHA-$_BUILD_ID \
         -var=cloudsql_instance=pgvector-instance-$SHORT_SHA-$_BUILD_ID \
-        -auto-approve -no-color -lock=false
+        -auto-approve -no-color 
         echo "pass" > /workspace/rag_tf_result.txt
 
         # Validate Ray: Make sure pods are running
-        kubectl wait --all pods -n rag-$SHORT_SHA-$_BUILD_ID --for=condition=Ready --timeout=1200s
-        kubectl port-forward -n rag-$SHORT_SHA-$_BUILD_ID service/ray-cluster-kuberay-head-svc 8265:8265 &
+        kubectl wait --for=condition=Ready pods -n rag-$SHORT_SHA-$_BUILD_ID -l 'component!=continuous-image-puller' --timeout=1200s
```
Comment (andrewsykim) [inline cloudbuild.yaml:231]: Why ` -l 'component!=continuous-image-puller'`?

**Pull request merged**